From ebfd1dd28133b4249de75d5864bf2a0d86885a9b Mon Sep 17 00:00:00 2001
From: Imran Rashid <irashid@cloudera.com>
Date: Wed, 22 Feb 2017 21:54:56 -0800
Subject: [PATCH 482/517] [SPARK-16554] Fix backport

---
 .../deploy/StandaloneDynamicAllocationSuite.scala  |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala b/core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala
index 005c6cb..14495c6 100644
--- a/core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala
+++ b/core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala
@@ -592,6 +592,8 @@ class StandaloneDynamicAllocationSuite
       val port = System.currentTimeMillis % 65536
       val endpointRef = mock(classOf[RpcEndpointRef])
       val mockAddress = mock(classOf[RpcAddress])
+      when(mockAddress.host).thenReturn("localhost")
+      when(mockAddress.hostPort).thenReturn(s"localhost:$port")
       when(endpointRef.address).thenReturn(mockAddress)
       val message = RegisterExecutor(id, endpointRef, s"localhost:$port", 10, Map.empty)
       val backend = sc.schedulerBackend.asInstanceOf[CoarseGrainedSchedulerBackend]
-- 
1.7.9.5

